File: src/main/java/org/cubexmc/commands/registrar/SuggestionProviders.java
package org.cubexmc.commands.registrar;

import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import org.bukkit.Bukkit;
import org.bukkit.command.CommandSender;
import org.cubexmc.RuleGems;
import org.cubexmc.commands.RuleGemsCommandActor;
import org.cubexmc.features.appoint.AppointFeature;
import org.cubexmc.model.GemDefinition;
import org.incendo.cloud.suggestion.Suggestion;
import org.incendo.cloud.suggestion.SuggestionProvider;

public final class SuggestionProviders {

    private final RuleGems plugin;

    public SuggestionProviders(RuleGems plugin) {
        this.plugin = plugin;
    }

    public SuggestionProvider<RuleGemsCommandActor> permSetSuggestions() {
        return SuggestionProvider.blocking((ctx, input) -> {
            AppointFeature feature = plugin.getFeatureManager() != null
                    ? plugin.getFeatureManager().getAppointFeature()
                    : null;
            if (feature == null) {
                return java.util.Collections.emptyList();
            }
            CommandSender sender = ctx.sender().sender();
            List<Suggestion> suggestions = new ArrayList<>();
            for (String key : feature.getAppointDefinitions().keySet()) {
                if (sender.hasPermission("rulegems.appoint." + key.toLowerCase(java.util.Locale.ROOT))
                        || sender.hasPermission("rulegems.appoint." + key)
                        || sender.hasPermission("rulegems.admin")) {
                    suggestions.add(Suggestion.suggestion(key));
                }
            }
            return suggestions;
        });
    }

    public SuggestionProvider<RuleGemsCommandActor> allPermSetSuggestions() {
        return SuggestionProvider.blocking((ctx, input) -> {
            AppointFeature feature = plugin.getFeatureManager() != null
                    ? plugin.getFeatureManager().getAppointFeature()
                    : null;
            if (feature == null) {
                return java.util.Collections.emptyList();
            }
            return feature.getAppointDefinitions().keySet().stream()
                    .map(Suggestion::suggestion)
                    .collect(Collectors.toList());
        });
    }

    public SuggestionProvider<RuleGemsCommandActor> gemKeySuggestions() {
        return SuggestionProvider.blocking((ctx, input) -> {
            Set<String> keys = new LinkedHashSet<>();
            for (GemDefinition definition : plugin.getGemParser().getGemDefinitions()) {
                if (definition == null || definition.getGemKey() == null || definition.getGemKey().isBlank()) {
                    continue;
                }
                keys.add(definition.getGemKey());
            }
            return keys.stream()
                    .sorted(String.CASE_INSENSITIVE_ORDER)
                    .map(Suggestion::suggestion)
                    .collect(Collectors.toList());
        });
    }

    public SuggestionProvider<RuleGemsCommandActor> onlinePlayerSuggestions() {
        return SuggestionProvider.blocking((ctx, input) -> Bukkit.getOnlinePlayers().stream()
                .map(p -> Suggestion.suggestion(p.getName()))
                .collect(Collectors.toList()));
    }
}

File: src/main/java/org/cubexmc/commands/registrar/SuggestionProviders.kt
package org.cubexmc.commands.registrar

import org.bukkit.Bukkit
import org.cubexmc.RuleGems
import org.cubexmc.commands.RuleGemsCommandActor
import org.incendo.cloud.suggestion.Suggestion
import org.incendo.cloud.suggestion.SuggestionProvider
import java.util.Locale

class SuggestionProviders(private val plugin: RuleGems) {
    fun permSetSuggestions(): SuggestionProvider<RuleGemsCommandActor> =
        SuggestionProvider.blocking { ctx, _ ->
            val feature = plugin.featureManager?.appointFeature
            if (feature == null) {
                return@blocking emptyList()
            }
            val sender = ctx.sender().sender()
            val suggestions = ArrayList<Suggestion>()
            for (key in feature.getAppointDefinitions().keys) {
                if (sender.hasPermission("rulegems.appoint." + key.lowercase(Locale.ROOT)) ||
                    sender.hasPermission("rulegems.appoint.$key") ||
                    sender.hasPermission("rulegems.admin")
                ) {
                    suggestions.add(Suggestion.suggestion(key))
                }
            }
            suggestions
        }

    fun allPermSetSuggestions(): SuggestionProvider<RuleGemsCommandActor> =
        SuggestionProvider.blocking { _, _ ->
            val feature = plugin.featureManager?.appointFeature
            if (feature == null) {
                return@blocking emptyList()
            }
            feature.getAppointDefinitions().keys.map { key -> Suggestion.suggestion(key) }
        }

    fun gemKeySuggestions(): SuggestionProvider<RuleGemsCommandActor> =
        SuggestionProvider.blocking { _, _ ->
            val keys = LinkedHashSet<String>()
            for (definition in plugin.gemParser.gemDefinitions) {
                val gemKey = definition?.gemKey
                if (gemKey.isNullOrBlank()) {
                    continue
                }
                keys.add(gemKey)
            }
            keys.sortedWith(String.CASE_INSENSITIVE_ORDER).map { key -> Suggestion.suggestion(key) }
        }

    fun onlinePlayerSuggestions(): SuggestionProvider<RuleGemsCommandActor> =
        SuggestionProvider.blocking { _, _ ->
            Bukkit.getOnlinePlayers().map { player -> Suggestion.suggestion(player.name) }
        }
}
